File: utils/clustering.py
import numpy as np
import pandas as pd
from sklearn.cluster import DBSCAN
from geopy.distance import geodesic


def cluster_by_gps(df, radius_meters=10, min_samples=2):
    """
    Cluster retailers based on GPS proximity using DBSCAN.
    Deduplicates by buyerid first - treating same buyerid as one unique retailer.

    Args:
        df: pandas DataFrame with currentLatitude and currentLongitude columns
        radius_meters: clustering radius in meters (default 10m)
        min_samples: minimum number of points to form a cluster

    Returns:
        pandas DataFrame: Original dataframe with added cluster_id column
    """
    if len(df) == 0:
        df['cluster_id'] = -1
        return df

    df = df.copy()

    # Store original dataframe for later merge
    df_original = df.copy()

    # Deduplicate by buyerid - keep the most recent meeting for each buyer
    # Also count total meetings per buyer
    df['meeting_count'] = df.groupby('buyerid')['buyerid'].transform('count')

    # Sort by date and keep the latest meeting for each buyer
    df = df.sort_values('meetingDate', ascending=False)
    df_unique = df.drop_duplicates(subset=['buyerid'], keep='first').copy()

    # Extract coordinates from unique buyers
    coords = df_unique[['currentLatitude', 'currentLongitude']].values

    # Convert to radians for haversine metric
    coords_rad = np.radians(coords)

    # Convert radius from meters to radians
    # Earth radius ≈ 6371 km
    earth_radius_km = 6371
    eps_rad = radius_meters / 1000 / earth_radius_km

    # Apply DBSCAN clustering on unique buyers
    clustering = DBSCAN(
        eps=eps_rad,
        min_samples=min_samples,
        metric='haversine'
    ).fit(coords_rad)

    # Add cluster labels to unique buyers dataframe
    df_unique['cluster_id'] = clustering.labels_

    # Merge cluster_id back to all meeting records based on buyerid
    df_with_clusters = df_original.merge(
        df_unique[['buyerid', 'cluster_id', 'meeting_count']],
        on='buyerid',
        how='left',
        suffixes=('', '_unique')
    )

    # Fill any missing cluster_ids with -1
    df_with_clusters['cluster_id'] = df_with_clusters['cluster_id'].fillna(-1).astype(int)
    df_with_clusters['meeting_count'] = df_with_clusters['meeting_count'].fillna(1).astype(int)

    return df_with_clusters


def calculate_cluster_stats(df):
    """
    Calculate statistics for each cluster based on unique buyers.

    Args:
        df: pandas DataFrame with cluster_id column

    Returns:
        pandas DataFrame: Cluster statistics
    """
    # Filter out noise points (cluster_id = -1)
    clustered = df[df['cluster_id'] >= 0].copy()

    if len(clustered) == 0:
        return pd.DataFrame()

    # Get unique buyers per cluster (deduplicate by buyerid)
    unique_buyers = clustered.drop_duplicates(subset=['buyerid', 'cluster_id'])

    # Group by cluster
    cluster_stats = unique_buyers.groupby('cluster_id').agg({
        'buyerid': 'count',  # Number of unique retailers in cluster
        'traderName': lambda x: x.nunique(),  # Number of unique traders
        'buyerName': lambda x: list(x),  # List of retailer names
        'buyerPhone': lambda x: list(x),  # List of phone numbers
        'currentLatitude': 'mean',  # Center latitude
        'currentLongitude': 'mean',  # Center longitude
    }).reset_index()

    cluster_stats.columns = [
        'cluster_id', 'retailer_count', 'trader_count',
        'retailer_names', 'phone_numbers',
        'center_lat', 'center_lon'
    ]

    return cluster_stats


def calculate_distance_matrix(df, cluster_id):
    """
    Calculate pairwise distances between unique retailers in a cluster.

    Args:
        df: pandas DataFrame
        cluster_id: ID of the cluster to analyze

    Returns:
        pandas DataFrame: Distance matrix
    """
    cluster_data = df[df['cluster_id'] == cluster_id].copy()

    # Get unique buyers only
    cluster_data = cluster_data.sort_values('meetingDate', ascending=False)
    cluster_data = cluster_data.drop_duplicates(subset=['buyerid'], keep='first')

    if len(cluster_data) < 2:
        return pd.DataFrame()

    # Get coordinates
    coords = cluster_data[['buyerName', 'currentLatitude', 'currentLongitude']].values

    # Calculate pairwise distances
    n = len(coords)
    distance_matrix = np.zeros((n, n))

    for i in range(n):
        for j in range(i + 1, n):
            lat1, lon1 = coords[i][1], coords[i][2]
            lat2, lon2 = coords[j][1], coords[j][2]
            distance = geodesic((lat1, lon1), (lat2, lon2)).meters
            distance_matrix[i, j] = distance
            distance_matrix[j, i] = distance

    # Create dataframe
    buyer_names = coords[:, 0]
    distance_df = pd.DataFrame(
        distance_matrix,
        index=buyer_names,
        columns=buyer_names
    )

    return distance_df


def get_cluster_details(df, cluster_id):
    """
    Get detailed information about a specific cluster (unique buyers only).

    Args:
        df: pandas DataFrame
        cluster_id: ID of the cluster to analyze

    Returns:
        pandas DataFrame: Detailed cluster information with unique buyers
    """
    cluster_data = df[df['cluster_id'] == cluster_id].copy()

    # Get unique buyers only (most recent meeting for each)
    cluster_data = cluster_data.sort_values('meetingDate', ascending=False)
    cluster_data = cluster_data.drop_duplicates(subset=['buyerid'], keep='first')

    # Select relevant columns
    detail_columns = [
        'buyerName', 'buyerid', 'buyerPhone', 'traderName', 'traderId',
        'Darkstore', 'meetingDate', 'currentLatitude', 'currentLongitude',
        'selfie', 'verificationDoc', 'remarks', 'meeting_count'
    ]

    available_columns = [col for col in detail_columns if col in cluster_data.columns]
    cluster_details = cluster_data[available_columns].copy()

    return cluster_details


def calculate_cluster_risk_score(df, cluster_stats, similarity_results):
    """
    Calculate risk score for each cluster based on multiple factors.

    Args:
        df: pandas DataFrame with all data
        cluster_stats: DataFrame with cluster statistics
        similarity_results: dict with similarity analysis results

    Returns:
        pandas DataFrame: cluster_stats with added risk_score column
    """
    cluster_stats = cluster_stats.copy()

    # Initialize risk score
    cluster_stats['risk_score'] = 0

    for idx, row in cluster_stats.iterrows():
        cluster_id = row['cluster_id']
        score = 0

        # Factor 1: Number of retailers (more = higher risk)
        retailer_count = row['retailer_count']
        if retailer_count >= 5:
            score += 30
        elif retailer_count >= 3:
            score += 20
        elif retailer_count >= 2:
            score += 10

        # Factor 2: Single trader managing all (higher risk)
        if row['trader_count'] == 1:
            score += 25

        # Factor 3: Name similarity (from similarity_results)
        cluster_key = f"cluster_{cluster_id}"
        if cluster_key in similarity_results:
            name_similarity_count = similarity_results[cluster_key].get('similar_name_pairs', 0)
            score += min(name_similarity_count * 15, 30)

        # Factor 4: Phone duplicates
        phone_list = row['phone_numbers']
        unique_phones = len(set(phone_list))
        if unique_phones < len(phone_list):
            score += 15

        cluster_stats.at[idx, 'risk_score'] = min(score, 100)

    # Categorize risk level
    cluster_stats['risk_level'] = cluster_stats['risk_score'].apply(
        lambda x: 'High' if x >= 60 else ('Medium' if x >= 30 else 'Low')
    )

    return cluster_stats

File: lib/clustering.ts
import type { Row, ClusterStat, SimilarityMap } from './types'

const EARTH_RADIUS_M = 6371000

export function haversineDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const toRad = (d: number) => (d * Math.PI) / 180
  const dLat = toRad(lat2 - lat1)
  const dLon = toRad(lon2 - lon1)
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2
  return 2 * EARTH_RADIUS_M * Math.asin(Math.sqrt(a))
}

function dbscan(
  points: { lat: number; lon: number }[],
  epsMeterss: number,
  minSamples: number
): number[] {
  const n = points.length
  const labels = new Array<number>(n).fill(-1)
  let clusterId = 0

  const neighbors = (idx: number): number[] => {
    const result: number[] = []
    for (let i = 0; i < n; i++) {
      if (i !== idx) {
        const d = haversineDistance(points[idx].lat, points[idx].lon, points[i].lat, points[i].lon)
        if (d <= epsMeterss) result.push(i)
      }
    }
    return result
  }

  for (let i = 0; i < n; i++) {
    if (labels[i] !== -1) continue
    const nbrs = neighbors(i)
    if (nbrs.length + 1 < minSamples) continue // noise for now

    labels[i] = clusterId
    const queue = [...nbrs]

    while (queue.length > 0) {
      const q = queue.shift()!
      if (labels[q] === -1) {
        labels[q] = clusterId
        const qNbrs = neighbors(q)
        if (qNbrs.length + 1 >= minSamples) {
          for (const nb of qNbrs) {
            if (labels[nb] === -1) queue.push(nb)
          }
        }
      } else if (labels[q] < 0) {
        labels[q] = clusterId
      }
    }
    clusterId++
  }

  return labels
}

export function clusterByGPS(rows: Row[], radiusMeters: number, minSamples: number): Row[] {
  if (rows.length === 0) return rows

  // Count meetings per buyerid+traderId pair
  const meetingCount: Record<string, number> = {}
  for (const r of rows) {
    const key = `${r.buyerid}::${r.traderId}`
    meetingCount[key] = (meetingCount[key] ?? 0) + 1
  }

  // Deduplicate by buyerid+traderId — keep most recent meeting
  const latestByPair: Record<string, Row> = {}
  for (const r of rows) {
    const key = `${r.buyerid}::${r.traderId}`
    const existing = latestByPair[key]
    if (!existing || r.meetingDate > existing.meetingDate) {
      latestByPair[key] = r
    }
  }
  const unique = Object.values(latestByPair)

  // Run DBSCAN on unique buyerid+traderId pairs
  const points = unique.map(r => ({ lat: r.currentLatitude, lon: r.currentLongitude }))
  const labels = dbscan(points, radiusMeters, minSamples)

  // Map buyerid+traderId → cluster_id
  const pairCluster: Record<string, number> = {}
  unique.forEach((r, i) => {
    const key = `${r.buyerid}::${r.traderId}`
    pairCluster[key] = labels[i]
  })

  // Merge back into all rows
  return rows.map(r => {
    const key = `${r.buyerid}::${r.traderId}`
    return {
      ...r,
      cluster_id: pairCluster[key] ?? -1,
      meeting_count: meetingCount[key] ?? 1,
    }
  })
}

export function calculateClusterStats(rows: Row[]): ClusterStat[] {
  const clustered = rows.filter(r => (r.cluster_id ?? -1) >= 0)
  if (clustered.length === 0) return []

  // Deduplicate by buyerid+traderId within each cluster
  const seen = new Set<string>()
  const uniqueRows: Row[] = []
  for (const r of clustered) {
    const key = `${r.cluster_id}:${r.buyerid}::${r.traderId}`
    if (!seen.has(key)) {
      seen.add(key)
      uniqueRows.push(r)
    }
  }

  const byCluster: Record<number, Row[]> = {}
  for (const r of uniqueRows) {
    const cid = r.cluster_id!
    if (!byCluster[cid]) byCluster[cid] = []
    byCluster[cid].push(r)
  }

  return Object.entries(byCluster).map(([cid, clusterRows]) => {
    const lats = clusterRows.map(r => r.currentLatitude)
    const lons = clusterRows.map(r => r.currentLongitude)
    const traders = new Set(clusterRows.map(r => r.traderName))

    return {
      cluster_id: parseInt(cid),
      retailer_count: clusterRows.length,
      trader_count: traders.size,
      retailer_names: clusterRows.map(r => r.buyerName),
      phone_numbers: clusterRows.map(r => r.buyerPhone),
      center_lat: lats.reduce((a, b) => a + b, 0) / lats.length,
      center_lon: lons.reduce((a, b) => a + b, 0) / lons.length,
      risk_score: 0,
      risk_level: 'Low',
    }
  })
}

export function calculateClusterRiskScore(
  clusterStats: ClusterStat[],
  similarityResults: SimilarityMap
): ClusterStat[] {
  return clusterStats.map(stat => {
    let score = 0

    // Factor 1: retailer count
    if (stat.retailer_count >= 5) score += 30
    else if (stat.retailer_count >= 3) score += 20
    else if (stat.retailer_count >= 2) score += 10

    // Factor 2: single trader
    if (stat.trader_count === 1) score += 25

    // Factor 3: name similarity
    const key = `cluster_${stat.cluster_id}`
    const sim = similarityResults[key]
    if (sim) {
      score += Math.min(sim.similar_name_pairs * 15, 30)
    }

    // Factor 4: phone duplicates
    const uniquePhones = new Set(stat.phone_numbers.filter(Boolean))
    if (uniquePhones.size < stat.phone_numbers.filter(Boolean).length) score += 15

    const risk_score = Math.min(score, 100)
    const risk_level: 'High' | 'Medium' | 'Low' =
      risk_score >= 60 ? 'High' : risk_score >= 30 ? 'Medium' : 'Low'

    return { ...stat, risk_score, risk_level }
  })
}

export function getClusterDetails(rows: Row[], clusterId: number): Row[] {
  const clusterRows = rows.filter(r => r.cluster_id === clusterId)
  // Latest meeting per buyerid+traderId pair
  const latest: Record<string, Row> = {}
  for (const r of clusterRows) {
    const key = `${r.buyerid}::${r.traderId}`
    if (!latest[key] || r.meetingDate > latest[key].meetingDate) {
      latest[key] = r
    }
  }
  return Object.values(latest)
}

export function getRetailerAllMeetings(rows: Row[], buyerid: string, traderId: string): Row[] {
  return rows
    .filter(r => r.buyerid === buyerid && r.traderId === traderId)
    .sort((a, b) => (a.meetingDate > b.meetingDate ? -1 : 1))
}
